pkg/tfshim/sdk-v2/provider.go

```diff
@@ -222,3 +224,149 @@ func (p v2Provider) IsSet(_ context.Context, v interface{}) ([]interface{}, bool
+	return &tfSchemaMarshaller{
+		Type:                    s.Type,
+		ConfigMode:              s.ConfigMode,
+		Required:                s.Required,
+		Optional:                s.Optional,
+		Computed:                s.Computed,
+		ForceNew:                s.ForceNew,
+		DiffSuppressFuncDefined: s.DiffSuppressFunc != nil,
+		DiffSuppressOnRefresh:   s.DiffSuppressOnRefresh,
+		Default:                 s.Default,
+		DefaultFuncDefined:      s.DefaultFunc != nil,
+		Description:             s.Description,
+		InputDefault:            s.InputDefault,
+		StateFuncDefined:        s.StateFunc != nil,
+		Elem:                    convertTFSchemaOrResource(s.Elem),
+		MaxItems:                s.MaxItems,
+		MinItems:                s.MinItems,
+		SetDefined:              s.Set != nil,
+		ComputedWhen:            s.ComputedWhen, //nolint:staticcheck
+		ConflictsWith:           s.ConflictsWith,
+		ExactlyOneOf:            s.ExactlyOneOf,
+		AtLeastOneOf:            s.AtLeastOneOf,
+		RequiredWith:            s.RequiredWith,
+		Deprecated:              s.Deprecated,
+		ValidateFuncDefined:     s.ValidateFunc != nil,
+		ValidateDiagFuncDefined: s.ValidateDiagFunc != nil,
+		Sensitive:               s.Sensitive,
+	}
+}
+
+func convertTfResource(r *schema.Resource) *tfResourceMarshaller {
+	tfResource := &tfResourceMarshaller{
+		Schema:                            make(map[string]*tfSchemaMarshaller),
+		SchemaVersion:                     r.SchemaVersion,
+		MigrateStateDefined:               r.MigrateState != nil, //nolint:staticcheck
+		StateUpgradersLen:                 len(r.StateUpgraders),
+		CustomizeDiffDefined:              r.CustomizeDiff != nil,
+		DeprecationMessage:                r.DeprecationMessage,
+		Description:                       r.Description,
+		UseJSONNumber:                     r.UseJSONNumber,
+		EnableLegacyTypeSystemApplyErrors: r.EnableLegacyTypeSystemApplyErrors,
+		EnableLegacyTypeSystemPlanErrors:  r.EnableLegacyTypeSystemPlanErrors,
+	}
+
+	for k, v := range r.Schema {
+		tfResource.Schema[k] = convertTFSchema(v)
+	}
+
+	return tfResource
+}
+
+func convertTfProv(p *schema.Provider) *tfProvMarshaller {
+	tfProv := &tfProvMarshaller{
+		Schema:               make(map[string]*tfSchemaMarshaller),
+		ResourcesMap:         make(map[string]*tfResourceMarshaller),
+		DataSourcesMap:       make(map[string]*tfResourceMarshaller),
+		ConfigureFuncDefined: p.ConfigureFunc != nil, //nolint:staticcheck
+		TerraformVersion:     p.TerraformVersion,
+	}
+
+	for k, v := range p.Schema {
+		tfProv.Schema[k] = convertTFSchema(v)
+	}
+
+	for k, v := range p.ResourcesMap {
+		tfProv.ResourcesMap[k] = convertTfResource(v)
+	}
+
+	for k, v := range p.DataSourcesMap {
+		tfProv.DataSourcesMap[k] = convertTfResource(v)
+	}
+
+	return tfProv
+}
+
+func (p v2Provider) DetailedSchemaDump() []byte {
+	prov := convertTfProv(p.tf)
+	sch, err := json.Marshal(prov)
+	contract.AssertNoErrorf(err, "failed to marshal schema")
+	return sch
+}
```

Comment:
You can get 70% of the information you need here for a lot less work:
```suggestion
func (p v2Provider) DetailedSchemaDump() []byte {
	prov, err := p.tf.GRPCProvider().GetProviderSchema(context.Background(), nil)
	sch, err := json.Marshal(prov)
	contract.AssertNoErrorf(err, "failed to marshal schema")
	return sch
}
```

Unfortunately, I think we still need the manual dump. A lot of the value we will get are from fields like `"DiffSuppressFuncDefined"` and `"DiffSuppressOnRefresh"` which are not available in TF's schema.

Reply:
The fact that a lot of the fields are not present there is the reason we are doing this.